core/src/main/java/aima/core/search/basic/uninformed/BidirectionalSearch.java

```diff
@@ -0,0 +1,127 @@
+        exploredBack.add(revNode.state());
+        Node<A, S> solutionNode;
+        while (!front.isEmpty() || !back.isEmpty()) {
+
+            if (front.isEmpty() || back.isEmpty()) {
+                return failure();
+            }
+            // Existence of path is checked from both ends of the problem.
+            solutionNode = pathExistsBidirectional(front, exploredFront, exploredBack, problem);
+            if (solutionNode != null) {
+                List<A> sol = solution(solutionNode);
+                Collections.reverse(sol);
+                return sol;
+            }
+            solutionNode = pathExistsBidirectional(back, exploredBack, exploredFront, problem);
+            if (solutionNode != null) {
+                List<A> sol = solution(solutionNode);
+                Collections.reverse(sol);
+                return sol;
+            }
+        }
+        return failure();
+    }
+
+    public Node<A, S> pathExistsBidirectional(Queue<Node<A, S>> queue, Set<S> exploredFront, Set<S> exploredBack, Problem<A, S> problem) {
+        if (!queue.isEmpty()) {
+            Node<A, S> next = queue.remove();
+            for (A action : problem.actions(next.state())) {
+                Node<A, S> child = newChildNode(problem, next, action);
+                if (exploredBack.contains(child.state())) {
+                    return child;
+                }
+                if (!(exploredFront.contains(child.state()) || containsState(queue, child.state()))) {
+                    if (problem.isGoalState(child.state())) {
+                        return child;
+                    }
+                    exploredFront.add(child.state());
+                    queue.add(child);
+                }
+            }
+        }
+        return null;
+    }
+
+    //Supporting Code
+    protected NodeFactory<A, S> nodeFactory = new BasicNodeFactory<>();
+    protected SearchController<A, S> searchController = new BasicSearchController<A, S>();
+
+    public Node<A, S> newRootNode(S initialState, double pathCost) {
+        return nodeFactory.newRootNode(initialState, pathCost);
+    }
+
+    public Node<A, S> newChildNode(Problem<A, S> problem, Node<A, S> node, A action) {
+        return nodeFactory.newChildNode(problem, node, action);
+    }
+
+    public List<A> solution(Node<A, S> node) {
+        return searchController.solution(node);
+    }
+
+    public List<A> failure() {
+        return searchController.failure();
+    }
+
+    public Queue<Node<A, S>> newFIFOQueue(Node<A, S> initialNode) {
+        Queue<Node<A, S>> frontier = new LinkedList<>();
+        frontier.add(initialNode);
+        return frontier;
+    }
+
+    public Set<S> newExploredSet() {
+        return new HashSet<>();
+    }
+
+    public BidirectionalSearch() {
+
+    }
+
+    public boolean containsState(Queue<Node<A, S>> frontier, S state) {
+        return frontier.stream().anyMatch(frontierNode -> frontierNode.state().equals(state));
+    }
```

Comment:
May want to consider a more efficient way of doing this. Have a look at BasicFrontierQueue, it will let you do a containment check much faster.